File: src-hs/Reduction.hs
module Reduction (
    Var(Var),
    Literal(Lit, NLit),
    Clause,
    CNF,

    Edges,
    Graph(Graph),

    Color(R, G, B),
    Coloring,
    Solution,
    SatSolver,

    generateCNF,
    recoverAnswer,
    solve
) where
    
import Control.Monad
import Control.Applicative

type Edges v = [(v, v)]
data Graph v = Graph { getVertices :: [v], getEdges :: Edges v }

data Var v = Var v Color
    deriving (Show, Eq)

data Literal v  = Lit (Var v) | NLit (Var v)
type Clause v   = [Literal v]
type CNF v      = [Clause v]
type Solution v = [(Var v, Bool)]

data Color = R | G | B
    deriving (Show, Enum, Eq)

type Coloring = [Color]

type SatSolver v = CNF v -> Maybe (Solution v)

solve :: Eq v => SatSolver v -> Graph v -> Maybe Coloring
solve solver graph = recoverAnswer graph $ solver $ generateCNF graph

recoverAnswer :: Eq v => Graph v -> Maybe (Solution v) -> Maybe Coloring
recoverAnswer _ Nothing = Nothing
recoverAnswer graph (Just xs) =
    mapM f (getVertices graph)
    where f v = do
            let col c = do
                val <- lookup (Var v c) xs
                guard val
                return c
            col R <|> col G <|> col B

generateCNF :: Graph v -> CNF v
generateCNF g =
    vertexHasColor g ++
    verticesColorUniqueness g ++
    adjVerticesHaveDiffColors g

{- Clauses for condition: each vertex has at least one color -}
vertexHasColor :: Graph v -> CNF v
vertexHasColor g = do
    v <- getVertices g
    return [ Lit $ Var v c | c <- [R, G, B] ]

{- Clauses for condition: no more than one color is assigned to each vertex -}
verticesColorUniqueness :: Graph v -> CNF v
verticesColorUniqueness g = do
    v <- getVertices g
    let diffCols c1 c2 = [ NLit $ Var v c1, NLit $ Var v c2 ]
    [ diffCols R B, diffCols B G, diffCols R G ]

{- Clauses for condition: adjacent vertices have different colors -}
adjVerticesHaveDiffColors :: Graph v -> CNF v
adjVerticesHaveDiffColors g = do
    (u, v) <- getEdges g
    [ [ NLit $ Var u c, NLit $ Var v c ] | c <- [R, G, B] ]

File: src-ml/Reduction.ml
module Graph = struct
  type 'v edge = 'v * 'v
  type 'v t = { vertices : 'v list; edges : 'v edge list }
end

module Color = struct
  type t = R | G | B
  type coloring = t list
end

module CNF = struct
  type 'v var = Var of 'v * Color.t
  type 'v literal = Lit of 'v var | NLit of 'v var
  type 'v clause = 'v literal list
  type 'v t = 'v clause list
  type 'v solution = 'v literal list
  type 'v solver = 'v t -> 'v solution option
end

(* Clauses for condition: each vertex has at least one color *)
let vertex_has_color (g : 'v Graph.t) : 'v CNF.t =
  let rec helper (vs : 'v list) =
    match vs with
    | [] -> []
    | v :: xs ->
        let lit c = CNF.Lit (CNF.Var (v, c)) in
        let cl = [ lit Color.R; lit Color.G; lit Color.B ] in
        cl :: helper xs
  in
  helper Graph.(g.vertices)

(* Clauses for condition: no more than one color is assigned to each vertex *)
let vertices_color_uniqueness (g : 'v Graph.t) : 'v CNF.t =
  let rec helper (vs : 'v list) =
    match vs with
    | [] -> []
    | v :: xs ->
        let nlit c = CNF.NLit (CNF.Var (v, c)) in
        let diff_cols c1 c2 = [ nlit c1; nlit c2 ] in
        let cl =
          [
            diff_cols Color.R Color.B;
            diff_cols Color.B Color.G;
            diff_cols Color.R Color.G;
          ]
        in
        cl @ helper xs
  in
  helper Graph.(g.vertices)

(* Clauses for condition: adjacent vertices have different colors *)
let adj_vertices_have_diff_colors (g : 'v Graph.t) : 'v CNF.t =
  let rec helper (edges : 'v Graph.edge list) =
    match edges with
    | [] -> []
    | e :: xs ->
        let u = fst e in
        let v = snd e in
        let nlit v c = CNF.Lit (CNF.Var (v, c)) in
        let cl1 = [ nlit u Color.R; nlit v Color.G ] in
        let cl2 = [ nlit u Color.G; nlit v Color.B ] in
        let cl3 = [ nlit u Color.B; nlit v Color.R ] in
        [ cl1; cl2; cl3 ] @ helper xs
  in
  helper Graph.(g.edges)

let generate_cnf (g : 'v Graph.t) : 'v CNF.t =
  vertex_has_color g
  @ vertices_color_uniqueness g
  @ adj_vertices_have_diff_colors g

let recover_answer (g : 'v Graph.t) (sol : 'v CNF.solution option) :
    Color.coloring option =
  match (g, sol) with
  | _, None -> None
  | graph, Some xs ->
      let rec helper (vs : 'v list) =
        match vs with
        | [] -> []
        | v :: xs ->
            let el =
              match v with
              | CNF.NLit v -> []
              | CNF.Lit (CNF.Var (v, c)) -> [ c ]
            in
            el @ helper xs
      in
      Some (helper Graph.(g.vertices))

let solve (sat_sol : 'v CNF.solver) (g : 'v Graph.t) : Color.coloring option =
  recover_answer g @@ sat_sol @@ generate_cnf g
